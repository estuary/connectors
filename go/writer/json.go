package writer

import (
	"compress/flate"
	"fmt"
	"io"

	"github.com/estuary/connectors/go/encrow"
	"github.com/klauspost/pgzip"
	"github.com/segmentio/encoding/json"
)

const (
	// JSON generally compresses well at minimum compression levels. Higher levels of compression
	// will usually take a lot more CPU while not providing much space savings.
	jsonCompressionlevel = flate.BestSpeed

	// Snowflake docs (https://docs.snowflake.com/en/sql-reference/sql/put#usage-notes) recommend
	// data file sizes be in the range of 100-250MB for compressed data. Redshift docs
	// (https://docs.aws.amazon.com/redshift/latest/dg/t_splitting-data-files.html) suggest files
	// for compressed data be in the range of 1MB to 1GB, and all be about the same size. Bigquery
	// docs don't mention anything other than the files must be less than 4GB. So, a 250MB file size
	// seems reasonable to use across the board given the current materializations that use this.
	DefaultJsonFileSizeLimit = 250 * 1024 * 1024
)

type jsonConfig struct {
	disableCompression bool
	skipNulls          bool
}

type JsonOption func(*jsonConfig)

func WithJsonDisableCompression() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.disableCompression = true
	}
}

func WithJsonSkipNulls() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.skipNulls = true
	}
}

type JsonWriter struct {
	w     io.Writer // will be set to `gz` for compressed writes or `cwc` if compression is disabled
	cwc   *countingWriteCloser
	gz    *pgzip.Writer
	shape *encrow.Shape
	buf   []byte
}

// NewJsonWriter creates a JsonWriter from w. w is closed when JsonWriter is closed. If `fields`
// is nil, values will be encoded as a JSON array rather than as an object.
func NewJsonWriter(w io.WriteCloser, fields []string, opts ...JsonOption) *JsonWriter {
	var cfg jsonConfig
	for _, o := range opts {
		o(&cfg)
	}

	jw := &JsonWriter{
		cwc: &countingWriteCloser{w: w},
	}

	if !cfg.disableCompression {
		gz, err := pgzip.NewWriterLevel(jw.cwc, jsonCompressionlevel)
		if err != nil {
			// Only possible if compressionLevel is not valid.
			panic("invalid compression level for gzip.NewWriterLevel")
		}
		jw.gz = gz
		jw.w = gz
	} else {
		jw.w = jw.cwc
	}

	if fields != nil {
		jw.shape = encrow.NewShape(fields)
		// Setting TrustRawMessage here prevents unnecessary validation of pre-serialized JSON
		// received from the runtime, which we can assume to be valid (flow_document for example).
		// Note that we are also not setting SortMapKeys or EscapeHTML: Sorting keys is not needed
		// because encrow.Shape already sorts the top-level keys and any object values are already
		// serialized as JSON, and escaping HTML is not desired so as to avoid escaping values like
		// <, >, &, etc. if they are present in the materialized collection's data.
		jw.shape.SetFlags(json.TrustRawMessage)
		jw.shape.SetSkipNulls(cfg.skipNulls)
	}

	return jw
}

func (w *JsonWriter) Write(vals []any) (err error) {
	w.buf = w.buf[:0]
	if w.shape == nil {
		// Serialize as a JSON array of values.
		w.buf = append(w.buf, '[')
		for idx, v := range vals {
			if w.buf, err = json.Append(w.buf, v, json.TrustRawMessage); err != nil {
				return fmt.Errorf("encoding JSON array value: %w", err)
			}
			if idx != len(vals)-1 {
				w.buf = append(w.buf, ',')
			}
		}
		w.buf = append(w.buf, ']')
	} else {
		// Serialize as a JSON object.
		if w.buf, err = w.shape.Encode(w.buf, vals); err != nil {
			return fmt.Errorf("encoding shape: %w", err)
		}
	}

	w.buf = append(w.buf, '\n')

	if _, err := w.w.Write(w.buf); err != nil {
		return fmt.Errorf("writing gzip bytes: %w", err)
	}

	return nil
}

func (w *JsonWriter) Written() int {
	return w.cwc.written
}

// Close closes the underlying gzip writer if compression is enabled, flushing its data and writing
// the GZIP footer. It also closes the underlying io.WriteCloser that was used to initialize the
// counting writer.
func (w *JsonWriter) Close() error {
	if w.gz != nil {
		if err := w.gz.Close(); err != nil {
			return fmt.Errorf("closing gzip writer: %w", err)
		}
	}

	if err := w.cwc.Close(); err != nil {
		return fmt.Errorf("closing counting writer: %w", err)
	}
	return nil
}
