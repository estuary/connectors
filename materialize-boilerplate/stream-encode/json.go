package stream_encode

import (
	"compress/flate"
	"fmt"
	"io"

	"github.com/estuary/connectors/go/encrow"
	"github.com/klauspost/compress/gzip"
	"github.com/segmentio/encoding/json"
)

const (
	// JSON generally compresses well at minimum compression levels. Higher levels of compression
	// will usually take a lot more CPU while not providing much space savings.
	jsonCompressionlevel = flate.BestSpeed

	// Snowflake docs (https://docs.snowflake.com/en/sql-reference/sql/put#usage-notes) recommend
	// data file sizes be in the range of 100-250MB for compressed data. Redshift docs
	// (https://docs.aws.amazon.com/redshift/latest/dg/t_splitting-data-files.html) suggest files
	// for compressed data be in the range of 1MB to 1GB, and all be about the same size. Bigquery
	// docs don't mention anything other than the files must be less than 4GB. So, a 250MB file size
	// seems reasonable to use across the board given the current materializations that use this.
	DefaultJsonFileSizeLimit = 250 * 1024 * 1024
)

type jsonConfig struct {
	disableCompression bool
}

type JsonOption func(*jsonConfig)

func WithJsonDisableCompression() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.disableCompression = true
	}
}

type JsonEncoder struct {
	w     io.Writer // will be set to `gz` for compressed writes or `cwc` if compression is disabled
	cwc   *countingWriteCloser
	gz    *gzip.Writer
	shape *encrow.Shape
	buf   []byte
}

// NewJsonEncoder creates a JsonEncoder from w. w is closed when JsonEncoder is closed. If `fields`
// is nil, values will be encoded as a JSON array rather than as an object.
func NewJsonEncoder(w io.WriteCloser, fields []string, opts ...JsonOption) *JsonEncoder {
	var cfg jsonConfig
	for _, o := range opts {
		o(&cfg)
	}

	enc := &JsonEncoder{
		cwc: &countingWriteCloser{w: w},
	}

	if !cfg.disableCompression {
		gz, err := gzip.NewWriterLevel(enc.cwc, jsonCompressionlevel)
		if err != nil {
			// Only possible if compressionLevel is not valid.
			panic("invalid compression level for gzip.NewWriterLevel")
		}
		enc.gz = gz
		enc.w = gz
	} else {
		enc.w = enc.cwc
	}

	if fields != nil {
		enc.shape = encrow.NewShape(fields)
		// Setting TrustRawMessage here prevents unnecessary validation of pre-serialized JSON
		// received from the runtime, which we can assume to be valid (flow_document for example).
		// Note that we are also not setting SortMapKeys or EscapeHTML: Sorting keys is not needed
		// because encrow.Shape already sorts the top-level keys and any object values are already
		// serialized as JSON, and escaping HTML is not desired so as to avoid escaping values like
		// <, >, &, etc. if they are present in the materialized collection's data.
		enc.shape.SetFlags(json.TrustRawMessage)
	}

	return enc
}

func (e *JsonEncoder) Encode(vals []any) (err error) {
	if e.shape == nil {
		// Serialize as a JSON array of values.
		e.buf = e.buf[:0]
		e.buf = append(e.buf, '[')
		for idx, v := range vals {
			if e.buf, err = json.Append(e.buf, v, json.TrustRawMessage); err != nil {
				return fmt.Errorf("encoding JSON array value: %w", err)
			}
			if idx != len(vals)-1 {
				e.buf = append(e.buf, ',')
			}
		}
		e.buf = append(e.buf, ']')
	} else {
		// Serialize as a JSON object.
		if e.buf, err = e.shape.Encode(e.buf, vals); err != nil {
			return fmt.Errorf("encoding shape: %w", err)
		}
	}

	e.buf = append(e.buf, '\n')

	if _, err := e.w.Write(e.buf); err != nil {
		return fmt.Errorf("writing gzip bytes: %w", err)
	}

	return nil
}

func (e *JsonEncoder) Written() int {
	return e.cwc.written
}

// Close closes the underlying gzip writer if compression is enabled, flushing its data and writing
// the GZIP footer. It also closes the underlying io.WriteCloser that was used to initialize the
// counting encoder.
func (e *JsonEncoder) Close() error {
	if e.gz != nil {
		if err := e.gz.Close(); err != nil {
			return fmt.Errorf("closing gzip writer: %w", err)
		}
	}

	if err := e.cwc.Close(); err != nil {
		return fmt.Errorf("closing counting writer: %w", err)
	}
	return nil
}
