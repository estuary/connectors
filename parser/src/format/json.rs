//! Parser for the json format. This will accept any stream of JSON values separated by whitespace.
//! It allows any amount of whitespace (including newlines) within and in between records.
use super::{Input, Output, ParseError, Parser};
use crate::ParseConfig;

struct JsonParser;

pub fn new_parser() -> Box<dyn Parser> {
    Box::new(JsonParser)
}

impl Parser for JsonParser {
    fn parse(&self, _config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
        let deser = serde_json::de::Deserializer::from_reader(content).into_iter();
        let wrapped = deser.map(|r| r.map_err(Into::into));
        Ok(Box::new(wrapped))
    }
}
