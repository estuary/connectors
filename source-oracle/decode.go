package main

import (
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"reflect"
	"strconv"
	"strings"
	"time"
	"unicode/utf16"

	"github.com/estuary/connectors/sqlcapture"
	"github.com/estuary/vitess/go/vt/sqlparser"
)

// decode SQL values extracted from the sql queries returned by logminer and parsed by sqlparser, into values
// to be used in the document
func decodeValue(expr sqlparser.Expr) (any, error) {
	switch v := expr.(type) {
	case *sqlparser.Literal:
		return v.Val, nil
	case *sqlparser.NullVal:
		return nil, nil
	case *sqlparser.OrExpr:
		// Oracle's string concatenation using pipes || is parsed as an OrExpr
		// UNISTR('literal string')
		var leftLiteral = v.Left.(*sqlparser.FuncExpr).Exprs[0].(*sqlparser.Literal).Val
		var left, err = decodeUnistr(leftLiteral)
		if err != nil {
			return nil, err
		}
		var rightLiteral = v.Right.(*sqlparser.FuncExpr).Exprs[0].(*sqlparser.Literal).Val
		right, err := decodeUnistr(rightLiteral)
		if err != nil {
			return nil, err
		}

		return left + right, nil
	case *sqlparser.FuncExpr:
		switch v.Name.String() {
		// Unicode strings (used for NVARCHAR2 literals with unicode characters)
		case "UNISTR":
			var val, err = decodeUnistr(v.Exprs[0].(*sqlparser.Literal).Val)
			if err != nil {
				return "", err
			}
			return val, nil
		// Used for DATE values
		case "TO_DATE":
			var val = v.Exprs[0].(*sqlparser.Literal).Val
			t, err := time.Parse(PARSE_DATE_FORMAT, val)
			if err != nil {
				return nil, fmt.Errorf("invalid date %q: %w", val, err)
			}
			return t.Format(OUT_DATE_FORMAT), nil
		case "TO_TIMESTAMP":
			var val = v.Exprs[0].(*sqlparser.Literal).Val

			// Timestamp has no fractional seconds, it ends up with a trailing dot
			if strings.HasSuffix(val, ".") {
				val += "0"
			}
			if t, err := time.Parse(PARSE_TS_FORMAT, val); err != nil {
				return nil, fmt.Errorf("invalid timestamp %q: %w", val, err)
			} else {
				return t.Format(OUT_TS_FORMAT), nil
			}
		case "TO_TIMESTAMP_TZ":
			var val = v.Exprs[0].(*sqlparser.Literal).Val
			// Timestamp has no fractional seconds, it ends up with a trailing dot
			if strings.HasSuffix(val, ".") {
				val += "0"
			}
			if t, err := time.ParseInLocation(PARSE_TSTZ_FORMAT, val, time.UTC); err != nil {
				// If the function is TO_TIMESTAMP_TZ but it has been output as TS format, then the value
				// is a local timestamp
				if tlocal, e := time.Parse(PARSE_TS_FORMAT, val); e == nil {
					return tlocal.Format(OUT_TS_FORMAT), nil
				}
				return nil, fmt.Errorf("invalid timestamptz %q: %w", val, err)
			} else {
				return t.UTC().Format(OUT_TSTZ_FORMAT), nil
			}
		case "TO_YMINTERVAL":
			var val = v.Exprs[0].(*sqlparser.Literal).Val
			return val, nil
		case "TO_DSINTERVAL":
			var val = v.Exprs[0].(*sqlparser.Literal).Val
			return val, nil
		case "HEXTORAW":
			var val = v.Exprs[0].(*sqlparser.Literal).Val
			var src = []byte(val)
			var hx = make([]byte, hex.DecodedLen(len(src)))
			if _, err := hex.Decode(hx, src); err != nil {
				return nil, fmt.Errorf("decoding hex %q: %w", val, err)
			}

			var b64 = make([]byte, base64.StdEncoding.EncodedLen(len(hx)))
			base64.StdEncoding.Encode(b64, hx)

			return string(b64), nil
		}
	}

	return nil, fmt.Errorf("unknown expression: %s %+v", reflect.TypeOf(expr).Name(), expr)
}

// decode the escaped unicode literals in UNISTR() functions
// the unicode literals are in the UCS-2 format
// see https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/UNISTR.html
func decodeUnistr(input string) (string, error) {
	var val string
	for i := 0; i < len(input); i++ {
		var c = input[i]
		// 92 = \ (backslash)
		if c == 92 {
			// two backslashes following each other is just an escaped backslash
			if len(input) > i+1 && input[i+1] == 92 {
				val += string(c)
				i += 1
			} else if len(input) > i+4 {
				var codeStrOne = input[i+1 : i+5]
				var codeOne, err = strconv.ParseInt(codeStrOne, 16, 32)
				if err != nil {
					return "", fmt.Errorf("parsing unicode point at %q (%d): %w", input, i+1, err)
				}
				if utf16.IsSurrogate(rune(codeOne)) {
					var codeStrTwo = input[i+6 : i+10]
					var codeTwo, err = strconv.ParseInt(codeStrTwo, 16, 32)
					if err != nil {
						return "", fmt.Errorf("parsing unicode point at %q (%d): %w", input, i+6, err)
					}
					i += 5

					val += string(utf16.DecodeRune(rune(codeOne), rune(codeTwo)))
				} else {
					val += string(rune(codeOne))
				}

				i += 4
			} else {
				val += string(c)
			}
		} else {
			val += string(c)
		}
	}

	return val, nil
}

func (s *replicationStream) decodeMessage(msg logminerMessage) (sqlcapture.DatabaseEvent, error) {
	var streamID = sqlcapture.JoinStreamID(msg.Owner, msg.TableName)

	var parser, err = sqlparser.New(sqlparser.Options{})
	if err != nil {
		return nil, err
	}
	ast, err := parser.Parse(msg.SQL)
	if err != nil {
		return nil, fmt.Errorf("parsing sql query %q: %w", msg.SQL, err)
	}
	undoAST, err := parser.Parse(msg.UndoSQL)
	if err != nil {
		return nil, fmt.Errorf("parsing undo sql query %q: %w", msg.UndoSQL, err)
	}
	var after, before map[string]any

	var op sqlcapture.ChangeOp
	switch msg.Op {
	case opInsert:
		after = make(map[string]any)
		op = sqlcapture.InsertOp
		insert, ok := ast.(*sqlparser.Insert)
		if !ok {
			return nil, fmt.Errorf("expected INSERT sql statement, instead got: %v", ast)
		}

		values, ok := insert.Rows.(sqlparser.Values)
		if !ok {
			return nil, fmt.Errorf("expected VALUES in INSERT statement, instead got: %v", insert.Rows)
		}
		// sql_redo field of logminer will have a single VALUES set for each insertion
		var vs = values[0]

		for i, col := range insert.Columns {
			var key = col.String()

			value, err := decodeValue(vs[i])
			if err != nil {
				return nil, fmt.Errorf("insert sql statement %q, key %q: %w", msg.SQL, key, err)
			}
			after[key] = value
		}

		// sql_redo of insert statements do not include a ROWID, but their SQL_UNDO which
		// is a delete statement does include a ROWID, so we extract the ROWID from the undo SQL
		del, ok := undoAST.(*sqlparser.Delete)
		if !ok {
			return nil, fmt.Errorf("expected DELETE undo sql statement, instead got: %v", undoAST)
		}

		err := sqlparser.VisitExpr(del.Where.Expr, func(node sqlparser.SQLNode) (kontinue bool, err error) {
			switch n := node.(type) {
			case *sqlparser.ComparisonExpr:
				var col = n.Left.(*sqlparser.ColName)
				var key = unquote(sqlparser.String(col.Name))
				if key == "ROWID" {
					var value = n.Right.(*sqlparser.Literal).Val
					after[key] = value

					return false, nil
				}
			}

			return true, nil
		})
		if err != nil {
			return nil, err
		}
	case opUpdate:
		before = make(map[string]any)
		after = make(map[string]any)
		op = sqlcapture.UpdateOp
		// construct `after` from the redo sql
		update, ok := ast.(*sqlparser.Update)
		if !ok {
			return nil, fmt.Errorf("expected UPDATE sql statement, instead got: %v", ast)
		}

		err := sqlparser.VisitExpr(update.Where.Expr, func(node sqlparser.SQLNode) (kontinue bool, err error) {
			switch n := node.(type) {
			case *sqlparser.ComparisonExpr:
				var col = n.Left.(*sqlparser.ColName)
				var key = unquote(sqlparser.String(col.Name))
				value, err := decodeValue(n.Right)
				if err != nil {
					return false, fmt.Errorf("update sql statement where clause %q, key %q: %w", msg.SQL, key, err)
				}
				after[key] = value
			}

			return true, nil
		})
		if err != nil {
			return nil, err
		}

		for _, expr := range update.Exprs {
			var key = unquote(sqlparser.String(expr.Name.Name))
			value, err := decodeValue(expr.Expr)
			if err != nil {
				return nil, fmt.Errorf("update sql statement %q, key %q: %w", msg.SQL, key, err)
			}
			after[key] = value
		}

		// construct `before` from the undo sql
		undo, ok := undoAST.(*sqlparser.Update)
		if !ok {
			return nil, fmt.Errorf("expected UPDATE sql statement, instead got: %v", ast)
		}
		err = sqlparser.VisitExpr(undo.Where.Expr, func(node sqlparser.SQLNode) (kontinue bool, err error) {
			switch n := node.(type) {
			case *sqlparser.ComparisonExpr:
				var col = n.Left.(*sqlparser.ColName)
				var key = unquote(sqlparser.String(col.Name))
				value, err := decodeValue(n.Right)
				if err != nil {
					return false, fmt.Errorf("update sql undo statement where clause %q, key %q: %w", msg.UndoSQL, key, err)
				}
				before[key] = value
			}

			return true, nil
		})
		if err != nil {
			return nil, err
		}

		for _, expr := range undo.Exprs {
			var key = unquote(sqlparser.String(expr.Name.Name))
			value, err := decodeValue(expr.Expr)
			if err != nil {
				return nil, fmt.Errorf("update sql undo statement %q, key %q: %w", msg.UndoSQL, key, err)
			}
			before[key] = value
		}
	case opDelete:
		before = make(map[string]any)
		op = sqlcapture.DeleteOp
		del, ok := ast.(*sqlparser.Delete)
		if !ok {
			return nil, fmt.Errorf("expected DELETE sql statement, instead got: %v", ast)
		}
		err := sqlparser.VisitExpr(del.Where.Expr, func(node sqlparser.SQLNode) (kontinue bool, err error) {
			switch n := node.(type) {
			case *sqlparser.ComparisonExpr:
				var col = n.Left.(*sqlparser.ColName)
				var key = unquote(sqlparser.String(col.Name))
				value, err := decodeValue(n.Right)
				if err != nil {
					return false, fmt.Errorf("delete sql statement %q, key %q: %w", msg.SQL, key, err)
				}
				before[key] = value
			}

			return true, nil
		})
		if err != nil {
			return nil, err
		}
	default:
		return nil, fmt.Errorf("unexpected operation code %d", msg.Op)
	}

	discovery, ok := s.tables.discovery[streamID]
	if !ok {
		return nil, fmt.Errorf("unknown discovery info for stream %q", streamID)
	}

	var rowid string
	if after != nil {
		rowid = after["ROWID"].(string)
		if err := translateRecordFields(discovery, after); err != nil {
			return nil, fmt.Errorf("error translating 'after' tuple: %w", err)
		}
	}

	if before != nil {
		rowid = before["ROWID"].(string)
		if err := translateRecordFields(discovery, before); err != nil {
			return nil, fmt.Errorf("error translating 'before' tuple: %w", err)
		}
	}

	keyColumns, ok := s.keyColumns(streamID)
	if !ok {
		return nil, fmt.Errorf("unknown key columns for stream %q", streamID)
	}

	var columnTypes = make(map[string]oracleColumnType)
	for name, column := range discovery.Columns {
		columnTypes[name] = column.DataType.(oracleColumnType)
	}

	var rowKey []byte
	if op == sqlcapture.InsertOp || op == sqlcapture.UpdateOp {
		rowKey, err = sqlcapture.EncodeRowKey(keyColumns, after, columnTypes, encodeKeyFDB)
	} else {
		rowKey, err = sqlcapture.EncodeRowKey(keyColumns, before, columnTypes, encodeKeyFDB)
	}
	if err != nil {
		return nil, fmt.Errorf("error encoding row key for %q: %w", streamID, err)
	}
	delete(after, "ROWID")
	delete(before, "ROWID")

	var sourceInfo = &oracleSource{
		SourceCommon: sqlcapture.SourceCommon{
			Millis:   msg.Timestamp,
			Schema:   msg.Owner,
			Snapshot: false,
			Table:    msg.TableName,
		},
		SCN:   msg.SCN,
		RowID: rowid,
	}

	var event = &sqlcapture.ChangeEvent{
		Operation: op,
		RowKey:    rowKey,
		Source:    sourceInfo,
		Before:    before,
		After:     after,
	}

	return event, nil
}
