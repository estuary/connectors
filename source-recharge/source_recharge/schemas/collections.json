{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "required": ["id"],
  "properties": {
    "id": {
      "type": ["integer"]
    },
    "name": {
      "type": ["null", "string"]
    },
    "created_at": {
      "type": ["null", "string"],
      "format": "date-time"
    },
    "updated_at": {
      "type": ["null", "string"],
      "format": "date-time"
    }
  }
}
