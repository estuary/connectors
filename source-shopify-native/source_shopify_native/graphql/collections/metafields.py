from datetime import datetime
from logging import Logger
from typing import AsyncGenerator
from ..metafields import MetafieldsResource

from source_shopify_native.models import SortKey


class CustomCollectionMetafields(MetafieldsResource):
    NAME = "custom_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return CustomCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:custom",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, CustomCollectionMetafields.PARENT_ID_KEY
        )


class SmartCollectionMetafields(MetafieldsResource):
    NAME = "smart_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return SmartCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:smart",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, SmartCollectionMetafields.PARENT_ID_KEY
        )
