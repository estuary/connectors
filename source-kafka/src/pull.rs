use crate::{
    configuration::{EndpointConfig, FlowConsumerContext, Resource},
    write_capture_response,
};
use anyhow::Result;
use hex::decode;
use highway::{HighwayHash, HighwayHasher, Key};
use lazy_static::lazy_static;
use proto_flow::{
    capture::{
        request::Open,
        response::{self, Checkpoint},
        Response,
    },
    flow::{capture_spec::Binding, ConnectorState, RangeSpec},
};
use rdkafka::{
    consumer::{Consumer, StreamConsumer},
    metadata::MetadataPartition,
    Message, Offset, TopicPartitionList,
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use tokio::io::{self};

#[derive(Debug, Deserialize, Serialize, Default)]
struct CaptureState {
    #[serde(rename = "bindingStateV1")]
    resources: HashMap<String, ResourceState>,
}

impl CaptureState {
    fn state_slice(state_key: &str, partition: i32, offset: i64) -> Self {
        let mut partitions = HashMap::new();
        partitions.insert(partition, offset);
        let mut resources = HashMap::new();
        resources.insert(state_key.to_string(), ResourceState { partitions });
        Self { resources }
    }
}

#[derive(Debug, Deserialize, Serialize, Default)]
struct ResourceState {
    partitions: HashMap<i32, i64>,
}

struct BindingInfo {
    binding_index: u32,
    state_key: String,
}

pub async fn do_pull(req: Open, mut stdout: io::Stdout) -> Result<()> {
    let spec = req.capture.expect("open must contain a capture spec");

    let state = if req.state_json == "{}" {
        CaptureState::default()
    } else {
        serde_json::from_str(&req.state_json)?
    };

    let config: EndpointConfig = serde_json::from_str(&spec.config_json)?;
    let mut consumer = config.to_consumer().await?;

    let topics_to_bindings =
        setup_consumer(&mut consumer, state, &spec.bindings, &req.range).await?;

    loop {
        let msg = consumer.recv().await?;

        let binding_info = topics_to_bindings
            .get(msg.topic())
            .expect("got a message for an unknown binding");

        let mut bytes = msg.payload().expect("message must contain a payload");
        // Strip a Confluent Schema Registry magic byte and schema ID.
        if bytes.starts_with(&[0]) && bytes.len() >= 5 {
            bytes = &bytes[5..];
        }

        let meta = json!({
            "topic": msg.topic(),
            "partition": msg.partition(),
            "offset": msg.offset()
        });

        let mut value: serde_json::Value = serde_json::from_slice(bytes)?;
        value
            .as_object_mut()
            .unwrap()
            .insert("_meta".to_string(), meta);

        let message = response::Captured {
            binding: binding_info.binding_index,
            doc_json: serde_json::to_string(&value)?,
        };

        let checkpoint =
            CaptureState::state_slice(&binding_info.state_key, msg.partition(), msg.offset());

        write_capture_response(
            Response {
                captured: Some(message),
                ..Default::default()
            },
            &mut stdout,
        )
        .await?;

        write_capture_response(
            Response {
                checkpoint: Some(Checkpoint {
                    state: Some(ConnectorState {
                        updated_json: serde_json::to_string(&checkpoint)?,
                        merge_patch: true,
                    }),
                }),
                ..Default::default()
            },
            &mut stdout,
        )
        .await?;
    }
}

async fn setup_consumer(
    consumer: &mut StreamConsumer<FlowConsumerContext>,
    state: CaptureState,
    bindings: &[Binding],
    range: &Option<RangeSpec>,
) -> Result<HashMap<String, BindingInfo>> {
    let meta = consumer.fetch_metadata(None, None)?;

    let extant_partitions: HashMap<String, &[MetadataPartition]> = meta
        .topics()
        .iter()
        .map(|t| (t.name().to_string(), t.partitions()))
        .collect();

    let mut topics_to_bindings: HashMap<String, BindingInfo> = HashMap::new();
    let mut topic_partition_list = TopicPartitionList::new();

    for (idx, binding) in bindings.iter().enumerate() {
        let res: Resource = serde_json::from_str(&binding.resource_config_json)?;

        let state_key = &binding.state_key;
        let topic = &res.topic;

        let default_state = ResourceState::default();
        let resource_state = state.resources.get(state_key).unwrap_or(&default_state);

        let partition_info = extant_partitions
            .get(topic)
            .expect("expected configured topic to exist");

        for partition in partition_info.iter() {
            let partition = partition.id();
            if !responsible_for_partition(range, &res.topic, partition) {
                continue;
            }

            let offset = match resource_state.partitions.get(&partition) {
                Some(o) => Offset::Offset(*o + 1), // Don't read the same offset again.
                None => Offset::Beginning,
            };
            topic_partition_list.add_partition_offset(topic, partition, offset)?;
        }

        topics_to_bindings.insert(
            topic.to_string(),
            BindingInfo {
                binding_index: idx as u32,
                state_key: state_key.to_string(),
            },
        );
    }

    consumer.assign(&topic_partition_list)?;

    Ok(topics_to_bindings)
}

lazy_static! {
    // HIGHWAY_HASH_KEY is a fixed 32 bytes (as required by HighwayHash) read from /dev/random.
    // DO NOT MODIFY this value, as it is required to have consistent hash results.
    // This value is copied from the Go connector source-boilerplate.
    static ref HIGHWAY_HASH_KEY: Vec<u8> = {
        decode("332757d16f0fb1cf2d4f676f85e34c6a8b85aa58f42bb081449d8eb2e4ed529f")
            .expect("invalid hex string for HIGHWAY_HASH_KEY")
    };
}

fn bytes_to_key(key: &[u8]) -> Key {
    assert!(key.len() == 32, "The key must be exactly 32 bytes long.");

    Key([
        u64::from_le_bytes(key[0..8].try_into().unwrap()),
        u64::from_le_bytes(key[8..16].try_into().unwrap()),
        u64::from_le_bytes(key[16..24].try_into().unwrap()),
        u64::from_le_bytes(key[24..32].try_into().unwrap()),
    ])
}

fn responsible_for_partition(range: &Option<RangeSpec>, topic: &str, partition: i32) -> bool {
    let range = match range {
        None => return true,
        Some(r) => r,
    };

    let mut hasher = HighwayHasher::new(bytes_to_key(&HIGHWAY_HASH_KEY));
    hasher.append(topic.as_bytes());
    hasher.append(&partition.to_le_bytes());
    let hash = (hasher.finalize64() >> 32) as u32;

    hash >= range.key_begin && hash <= range.key_end
}
