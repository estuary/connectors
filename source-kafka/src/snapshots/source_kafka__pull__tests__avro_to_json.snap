---
source: src/pull.rs
expression: "avro_to_json(record.into(), &record_schema_parsed).unwrap()"
---
{
  "arrayField": [
    "first",
    "second"
  ],
  "boolField": true,
  "bytesField": "aGVsbG8=",
  "dateField": 123,
  "decimalField": "-32.442",
  "doubleField": 340282346638528860000000000000000000000.0,
  "doubleFieldNaN": "NaN",
  "doubleFieldNegInf": "-inf",
  "doubleFieldPosInf": "inf",
  "durationField": "P6M14DT20H30M",
  "enumField": "b",
  "fixedBytesField": "aGVsbG8=",
  "fixedDecimalField": "99.36",
  "floatField": 340282346638528860000000000000000000000.0,
  "floatFieldNaN": "NaN",
  "floatFieldNegInf": "-inf",
  "floatFieldPosInf": "inf",
  "intField": 2147483647,
  "localTimestampMicrosField": "2024-10-29T20:26:46Z",
  "localTimestampMillisField": "2024-10-29T20:26:46Z",
  "longField": 9223372036854775807,
  "mapField": {
    "key": "value"
  },
  "nestedRecordField": {
    "nestedLongField": 123,
    "nestedStringField": "nested string value"
  },
  "nullField": null,
  "nullableStringField": null,
  "stringField": "hello",
  "timeMicrosField": "20:30:00.000000",
  "timeMillisField": "20:30:00.000",
  "timestampMicrosField": "2024-10-29T20:26:46Z",
  "timestampMillisField": "2024-10-29T20:26:46Z",
  "uuidField": "00000000-0000-0000-0000-000000000000"
}
